clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -123,16 +123,28 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                 // enough memory to allocate one
                 while (accumulated < size) {
                     long startWaitNs = time.nanoseconds();
-                    boolean waitingTimeElapsed = !moreMemory.await(remainingTimeToBlockNs, TimeUnit.NANOSECONDS);
-                    long endWaitNs = time.nanoseconds();
-                    long timeNs = Math.max(0L, endWaitNs - startWaitNs);
+                    boolean waitingTimeElapsed = false;
+                    long endWaitNs;
+                    long timeNs;
```

Comment:
I think it would be better not to declare these two values here, but when they are defined. That will make it clearer that if we get past the try/catch block, these values cannot have been set yet (because the exception is rethrown). I mean something like:

```java
                    long startWaitNs = time.nanoseconds();
                    boolean waitingTimeElapsed = false;
                    try {
                        waitingTimeElapsed = !moreMemory.await(remainingTimeToBlockNs, TimeUnit.NANOSECONDS);
                    } catch (InterruptedException e) {
                        this.waiters.remove(moreMemory);
                        long endWaitNs = time.nanoseconds();
                        long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                        this.waitTime.record(timeNs, time.milliseconds());
                        throw e;
                    }
                    long endWaitNs = time.nanoseconds();
                    long timeNs = Math.max(0L, endWaitNs - startWaitNs);
```

Does that make sense?